File: 2020/q2.cc
/* See q2.py for notes. This solution is fast enough to get full marks :) */

#include <algorithm>
#include <iostream>
#include <unordered_set>
#include <vector>

using namespace std;

vector<vector<int>> construct_map(int r, vector<int> plan) {
  vector<vector<int>> connections(r);
  unordered_set<int> chosen;
  unordered_set<int> all;

  for (int i = 0; i < r; i++) {
    all.insert(i);
  }

  while (plan.size() > 0) {
    int next = plan.back();
    int picked = -1;

    // cout << "next = " << next << endl;

    for (int i = 0; i < r; i++) {
      if (chosen.count(i) == 0 &&
          find(plan.begin(), plan.end(), i) == plan.end()) {
        picked = i;
        chosen.insert(i);
        break;
      }
    }

    if (picked == -1) {
      throw "trying to pick -1";
    }

    // cout << "connecting " << next << " and " << picked << endl;
    connections[next].push_back(picked);
    connections[picked].push_back(next);

    plan.pop_back();
  }

  std::vector<int> available;

  for (int i = 0; i < r; i++) {
    if (chosen.count(i) == 0) {
      available.push_back(i);
    }
  }

  int x = available.at(0);
  int y = available.at(1);

  // cout << "connecting " << x << " and " << y << endl;

  connections[x].push_back(y);
  connections[y].push_back(x);

  for (int i = 0; i < r; i++) {
    sort(connections[i].begin(), connections[i].end());
  }

  return connections;
}

int main() {
  string plan_s;
  vector<int> plan;
  int p, q;

  cin >> plan_s >> p >> q;

  for (int i = plan_s.length() - 1; i >= 0; i--) {
    // the plan is in reverse, so we can use the thing as a stack.
    plan.push_back(int(plan_s.at(i) - 'A'));
  }

  int r = plan.size() + 2;

  auto connections = construct_map(r, plan);

  for (int i = 0; i < r; i++) {
    for (int connection_i : connections[i]) {
      char connection = 'A' + char(connection_i);

      cout << connection;
    }

    cout << endl;
  }

  // now run the game
  int start_room = 0, move_count = 0;
  int current_room = start_room;

  std::vector<int> room_counts(r);
  std::vector<std::vector<int>> leave_counts(r);
  for (int i = 0; i < r; i++) {
    leave_counts[i].resize(r);
  }

  room_counts[start_room] = 1;

  while (move_count < q) {
    int visit_count = room_counts[current_room];
    int leave_through = -1;

    if (visit_count % 2 == 1) {
      // odd
      leave_through = connections[current_room][0];
    } else {
      auto &exits = connections[current_room];

      for (int i = 0; i < exits.size(); i++) {
        auto exit = exits[i];
        if (leave_counts[current_room][exit] % 2 == 1) {
          if (i == exits.size() - 1) {
            leave_through = exit;
          } else {
            leave_through = exits[i + 1];
          }
          break;
        }
      }
    }

    leave_counts[current_room][leave_through] += 1;
    room_counts[leave_through] += 1;
    current_room = leave_through;

    move_count++;

    if (move_count == p) cout << char('A' + current_room);
    if (move_count == q) cout << char('A' + current_room);
  }

  cout << endl;
  return 0;
}

File: 2020/q2.py
"""
Q2: Alpha Complex

Frustratingly this is too slow for two of the larger examples (test 6 and test
8). I think this is mostly just an issue with Python :(.

Still, only lost 6 marks for to slowness. 18/24.

2b: (i) A (ii) AAAA

2c: The first time we visit a room, all of the exits have been used an even
(zero) number of times. Each time we visit the room, we flip one exit from odd
to even. Add up all of the rooms (1 for odd, 0 for even). If the resulting
number is odd, the room number is even. If the resulting number is even, the
room number is odd. (3/4)

2d: There's probably a clever combinatorical way to calculate this, but I'll
brute force it. 8! = 40k, basically nothing.
"""

"""
Stage 1: construct the map.
"""


def construct_map(r, plan):
    all = set(i for i in range(r))
    connections = [set() for _ in range(r)]
    chosen = set()

    while len(plan) > 0:
        # pick first room alphabetically
        picked = None

        for i in range(r):
            if i not in chosen and i not in plan:
                picked = i
                chosen.add(picked)
                break

        # print("connecting %s and %s" % ((chr(ord("A") + picked), chr(ord("A") + plan[0]))))

        connections[picked].add(plan[0])
        connections[plan[0]].add(picked)

        plan = plan[1:]
        # print("plan = %s" % ("".join(chr(ord("A") + x) for x in plan)))

    unpicked = all - chosen
    x, y = tuple(unpicked)

    # print("connecting %s and %s" % ((chr(ord("A") + x), chr(ord("A") + y))))

    connections[x].add(y)
    connections[y].add(x)

    return connections


plan, p, q = input().split(" ")
p = int(p)
q = int(q)
plan = list(plan)
r = len(plan) + 2

# Rename rooms from letters to numbers. (A=0, B=1 etc.)
plan = [ord(c) - ord("A") for c in plan]
connections = construct_map(r, plan)

# Print the rooms connected to A

connections = [list(sorted(y)) for y in connections]

for i in range(r):
    connected_to_a = []
    for y in connections[i]:
        connected_to_a.append(chr(ord("A") + y))

    print("".join(connected_to_a))

start_room = 0  # A

room_counts = [0 for _ in range(r)]
room_counts[start_room] = 1

# all exists are zero
leave_counts = [[0 for _ in range(r)] for _ in range(r)]

current_room = start_room
move_count = 0

after_p = None
after_q = None

while move_count < q:
    visit_count = room_counts[current_room]

    leave_through = None

    if visit_count % 2 == 1:
        # odd
        # leave through the first exit alphabetically
        leave_through = connections[current_room][0]
    else:
        # even
        # find the first exit alphabetically that they have left through an
        # odd number of times

        exits = connections[current_room]

        for i, exit in enumerate(exits):
            if leave_counts[current_room][exit] % 2 == 1:
                if i == len(exits) - 1:
                    leave_through = exit
                    break
                else:
                    # not the last one, leave through the next one, alphabetically
                    leave_through = exits[i + 1]
                    break

    leave_counts[current_room][leave_through] += 1
    room_counts[leave_through] += 1
    current_room = leave_through

    move_count += 1

    if move_count == p:
        after_p = current_room
    if move_count == q:
        after_q = current_room

print("%s%s" % (chr(ord("A") + after_p), chr(ord("A") + after_q)))
